src/render_engine/gl3/gl3_render_engine.cpp

```diff
@@ -11,18 +12,52 @@
 
 namespace nova::renderer::rhi {
     Gl4NvRenderEngine::Gl4NvRenderEngine(NovaSettingsAccessManager& settings) : RenderEngine(&mallocator, settings) {
-
         window = std::make_unique<GlfwWindow>(settings.settings);
 
         gladLoadGLLoader(GlfwWindow::get_gl_proc_address);
 
+        save_device_info();
+
         swapchain = new Gl3Swapchain(settings.settings.max_in_flight_frames, window->get_window_size());
 
         set_initial_state();
     }
 
     Gl4NvRenderEngine::~Gl4NvRenderEngine() { delete swapchain; }
 
+    void Gl4NvRenderEngine::save_device_info() {
+        const GLubyte* raw_vendor_string = glGetString(GL_VENDOR);
+        std::string vendor_string(reinterpret_cast<const char*>(raw_vendor_string));
+        std::transform(vendor_string.begin(), vendor_string.end(), vendor_string.begin(), [](const char c) { return std::tolower(c); });
```

Comment:
ugh this is why c++ is dumb

Reply:
Yeah I had a whole rant about this at one point

Reply:
I remember seeing it lmao, I wish we could backport some of the rust iterator interface to c++.

Reply:
Whenever I get around to making a Nova Template Library it'll have something like Java Streams or C# LINQ or Rust iter()